src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  let currentState = { ...state };
+
+  for (let i = 0; i < actions.length; i++) {
+    if (actions[i].type === 'addProperties') {
+      currentState = { ...currentState, ...actions[i].extraData };
+
+      result.push(currentState);
+    }
+
+    if (actions[i].type === 'removeProperties') {
+      const newState = { ...currentState };
+
+      for (const property of actions[i].keysToRemove) {
+        delete newState[property];
```

Comment:
The `delete` operator is used correctly to remove properties. Ensure that `actions[i].keysToRemove` is always an array of strings to prevent runtime errors.